File: Past/20_Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dic = {"]":"[", "}":"{", ")":"("}
        
        for char in s:
            if char in dic.values():
                stack.append(char)
            elif char in dic.keys():
                if stack == [] or dic[char] != stack.pop():
                    return False
            
        return stack == []

File: Algorithms/20. Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const bracket = [];

  for (let char of s) {
    switch (char) {
      case '(': {
        bracket.push(char);
        break;
      }
      case '[': {
        bracket.push(char);
        break;
      }
      case '{': {
        bracket.push(char);
        break;
      }
      case ')': {
        if (bracket.pop() !== '(') {
          return false;
        }
        break;
      }
      case ']': {
        if (bracket.pop() !== '[') {
          return false;
        }
        break;
      }
      case '}': {
        if (bracket.pop() !== '{') {
          return false;
        }
        break;
      }
    }
  }

  if (bracket.length !== 0) return false;
  return true;
};

module.exports = isValid;
